pkg/segment/segment_test.go

```diff
@@ -0,0 +1,191 @@
+	body, server := mockServer()
+	defer server.Close()
+	defer close(body)
+	falseValue := false
+
+	cfg := &preference.PreferenceInfo{
+		Preference: preference.Preference{
+			OdoSettings: preference.OdoSettings{
+				ConsentTelemetry: &falseValue,
+			},
+		},
+	}
+
+	c, err := newCustomClient(cfg, filepath.Join(os.TempDir(), "telemetry"), server.URL)
+	//	assert err != nil
+	if err != nil {
+		t.Fatal(err)
+	}
+	// run a command, odo preference view
+	if err := c.Upload("odo preference view", time.Second, errors.New("an error occured")); err != nil {
+		t.Fatal(err)
+	}
+	if err := c.Close(); err != nil {
+		t.Fatal(err)
+	}
+
+	select {
+	case <-body:
+		t.Fatal("server should not receive data")
+	default:
+	}
+}
+
+func TestClientUploadWithConsent(t *testing.T) {
+	body, server := mockServer()
+	defer server.Close()
+	defer close(body)
+
+	trueValue := true
+
+	cfg := &preference.PreferenceInfo{
+		Preference: preference.Preference{
+			OdoSettings: preference.OdoSettings{
+				ConsentTelemetry: &trueValue,
+			},
+		},
+	}
+	tests := []struct {
+		testName string
+		err      error
+		success  bool
+		errType  string
+	}{
+		{
+			testName: "command ran successfully",
+			err:      nil,
+			success:  true,
+			errType:  "",
+		},
+		{
+			testName: "command failed",
+			err:      errors.New("some error occurred"),
+			success:  false,
+			errType:  "*errors.errorString",
+		},
+	}
+	for _, tt := range tests {
+		t.Log("Running test: ", tt.testName)
+		t.Run(tt.testName, func(t *testing.T) {
+			c, err := newCustomClient(cfg, filepath.Join(os.TempDir(), "telemetry"), server.URL)
+			if err != nil {
+				t.Error(err)
+			}
+			//run a command, odo preference view
+			if err := c.Upload("odo preference view", time.Second, tt.err); err != nil {
+				t.Error(err)
+			}
+			if err := c.Close(); err != nil {
+				t.Error(err)
+			}
+
+			select {
+			case x := <-body:
+				s := segmentResponse{}
+				if err := json.Unmarshal(x, &s); err != nil {
+					t.Error(err)
+				}
+				if s.Batch[0].Type != "identify" && s.Batch[1].Type != "track" {
+					t.Error("Missing Identify or Track information")
+				}
+				if s.Batch[0].Traits.OS != runtime.GOOS {
+					t.Error("OS does not match")
+				}
+				if !tt.success {
+					if s.Batch[1].Properties.Error != tt.err.Error() {
+						t.Error("Error does not match")
+					}
+				} else {
+					if s.Batch[1].Properties.Error != "" {
+						t.Error("Error does not match")
+					}
+				}
+				if s.Batch[1].Properties.Success != tt.success {
+					t.Error("Success does not match")
+				}
+				if s.Batch[1].Properties.ErrorType != tt.errType {
+					t.Error("Error Type does not match")
+				}
+				if !strings.Contains(s.Batch[1].Properties.Version, version.VERSION) {
+					t.Error("Odo version does not match")
+				}
+
+			default:
+				t.Error("Server should receive data")
+			}
+		})
+	}
+}
+
+func TestSetError(t *testing.T) {
+	err := errors.New("this is an error string")
```

Comment:
not sure whats the purpose of this?

Reply:
To check if the username, incase it appears in an error, is sanitized before being sent to Segment. I'll be adding more tests related to sanitization once we add more sanitization checks.

Reply:
The test first checks a proper error string that does not contain any PII and makes sure the error returned by `SetError` is the same as the error itself. Next, it passes an error with a username(PII) and makes sure the error returned by `SetError` is not the same as the error itself and that the username was sanitized.